rfcs/0116-artifact-service.md

```diff
@@ -0,0 +1,205 @@
+IP will occur to find the backing storage.  If the origin is not an IPv4 or
+IPv6 address and is identical to a set of known and mappable identifiers (e.g.
+`s3_us-west-2`), then that will be used to map the token to find the backing
+storage.
+
+### Creation
+
+```
+PUT /artifacts/:name
+PATCH /artifacts/:name
+```
+
+The `PUT` endpoint will take an origin and name.  It will return a `200` with a
+document that contains a list of generalised requests which could be run to
+complete the actual upload.  Please see the queue's current `blob` storage type
+for an example of this pattern.  The body of the request will contain metadata
+about the artifact in the shape:
+
+```javascript
+{
+  origin: '10.10.10.10' || 's3_us-west-2',
+  contentType: 'application/json',
+  contentLength: 1234,
+  contentSha256: 'abcdef1234',
+  transferLength: 1234,
+  transferSha256: 'abcdef1234',
+  contentEncdoing: 
+  expiration: new Date(),
+}
+```
+
+This endpoint will return a list of HTTP requests which must all be run.
+
+The `PATCH` endpoint will be sent without a request body and will be used to
+perform any post-upload actions.  These include things like the commit step on
+an S3 multipart upload.  This endpoint would be run by the Queue and not by the
+uploader, and the Queue would also need a complimentary method for workers to
+call when they complete the upload.  This is what is currently done with the
+`blob` storage type in the queue.
+
+### Deletetion
+
+```
+DELETE /artifacts/:name
+DELETE /caches/artifacts/:name
+```
+
+This service has at least one copy of each stored file.  Any copies above 
+
+Delete all stored copies of the named object if using first endpoint.  The
+second endpoint would attempt delete all of the cached copies in the various
+locations.  The cached copies are those copies which are created in other
+regions or services to provide more local copies of each object.  This cache
+purging must return a `202 Accepted` response and not `200/204` due to the
+nature of how caches work in the real world.
+
+### Retreival
+
+```
+GET /artifacts/:name[?max_time=30]
+```
+
+This endpoint returns a 302 redirect to the location of the object which is
+closest to the request's IP address.  If needed, this will initiate a copy from
+the original backing store into a different storage location.  It will wait up
+to a configurable amount of time before redirecting to a non-optimal copy.
+
+In order to support running in environments where requests are limted in time,
+the `max_time` query string option will specify roughly how many seconds the
+client is willing to wait for an artifact.  The default value should be based
+on the size of the artifact.  For example, it will wait 2s for each 100MB of
+artifact size.  While waiting, the service will issue intermediate redirects to
+itself to reduce the waiting time.
+
+Example inside heroku:
+```
+10:00:00 <-- GET /artifacts/my_file?max_time=90
+10:00:25 --> 302 /artifacts/my_file?max_time=65
```

Comment:
This is probably unwise as most clients stop following redirects after 5-7 hops. We can keep a TCP connection alive longer by writing whitespace in HTTP, or some other hack like that.

Okay, I'm not sure.. just saying there is something to think about here.

Reply:
I'm not sure where you're getting that 5-7 amount.  Clients can choose to follow more redirects if they want.  This API also doesn't force us a specific mechanism of waiting, since the user says "Hey, I'll wait 200 seconds" and we just do our best to stall them for 200s.

Just like outside of heroku, we cannot write to the body of a response before we know where the 302 should redirect to as headers would've had to have been fully sent by the time the first byte is written.  We could redirect to a waiter endpoint, but then that waiter endpoint would require *another* 302 every single time since we'd have to redirect back to the original endpoint.  And since this is all HTTP spec compliant stuff, there's nothing stopping us from changing between the two approaches whenever we choose.

By the time we write whitespace, we'd need to know if the request redirected to the original bucket or to the copy.

It's something to think about for sure, but I'm not too concerned.  Based on this totally, 100% legit, A-Grade [site](https://websiteadvantage.com.au/Google-Maximum-Redirects), browsers follow >=20 redirects, which is 10 minutes of waiting in heroku.  Curl does 50 by default on my mac.  `curl -LI https://httpbin.org/redirect/1000`

Reply:
Okay, it's possible I'm wrong.

I think I noticed a lot of clients in npm, golang that defaults to something like 5-7 redirects.
 * golang by default [follows 10 redirects](https://golang.org/pkg/net/http/#Client) (not bad).
 * npm `request` follows [10 by default](https://www.npmjs.com/package/request#requestoptions-callback).
 * npm `got` is also [10 redirects](https://www.npmjs.com/package/got#gotmaxredirectserror)
 * python requests [defaults to 30 redirects](http://docs.python-requests.org/en/master/api/#requests.Session.max_redirects)
 * npm `superagent` [defaults to 5 redirects](https://visionmedia.github.io/superagent/#following-redirects) (we already don't like this)
 * `wget` defaults to [20 redirects](https://www.gnu.org/software/wget/manual/wget.html) (could be build dependent)
 * `curl` defaults to [50 redirects](https://curl.haxx.se/docs/manpage.html#--max-redirs) if following redirects is enabled with `-L`

So maybe it's not such a crazy idea after all :)